File: experiments/source/ocaml/soft_threshold_reordered.ml
open Loom

let[@loom.entry] soft_threshold_reordered (threshold : float) (x : Tensor1.t) =
  let shrink xi =
    if xi < (0.0 -. threshold) then xi +. threshold
    else if threshold < xi then xi -. threshold
    else 0.0
  in
  Tensor1.map shrink x

File: experiments/source/python/soft_threshold_reordered.py
import loom


@loom.entry
def soft_threshold_reordered(threshold: float, x: loom.Tensor1) -> loom.Tensor1:
    def shrink(xi: float) -> float:
        return xi + threshold if xi < (0.0 - threshold) else xi - threshold if threshold < xi else 0.0

    return loom.Tensor1.map(shrink, x)
